File: backend/routes/platform/roles.py
########## Modules ##########
from fastapi import APIRouter, Request, Depends

from sqlalchemy import or_
from sqlalchemy.orm import Session

from db.database import get_db
from db.model import User_Role

from core.i18n import translate
from core.generator import get_uuid
from core.responses import custom_response
from core.db_management import add_db, update_db
from core.validators import read_json_body, validate_required_fields
from core.permissions import get_permissions, filter_existing_permissions, check_permissions

########## Variables ##########
router = APIRouter()

########## Get Roles - Dashboard ##########
@router.get("/")
async def roles_platform(request: Request, db: Session = Depends(get_db)):
    ### Variables ###
    lang = request.state.lang
    user = request.state.user
    roles = []

    ### Validation ###
    if user == None:
        return custom_response(status_code=400, message=translate(lang, "validation.require_auth"))
    
    ### Check permissions ###
    access, message = check_permissions(db, request, "platform.roles.read")
    
    if not access:
        return custom_response(status_code=400, message=message)
    
    ### Operations ###
    filters = [
        User_Role.platform_level == True,
    ]

    role_access, _ = check_permissions(db, request, "platform.users.role")

    if role_access:
        filters = [
            or_(
                User_Role.platform_level.is_(True),
                User_Role.hidden.is_(True)
            )
        ]

    roles_data = db.query(User_Role).filter(*filters).all()

    for role in roles_data:
        roles.append({
            "id": role.id,
            "name": role.name,
            "platform": role.platform_level,
            "date": role.date.strftime("%d %B %Y")
        })
        
    return custom_response(status_code=200, message=translate(lang, "platform.roles.get.success"), data={
        "roles": roles
    })

########## Get Permissions - Create Role ##########
@router.get("/get-permissions")
async def get_roles(request: Request, db: Session = Depends(get_db)):
    ### Variables ###
    lang = request.state.lang
    user = request.state.user
    permissions = []

    ### Validation ###
    if user == None:
        return custom_response(status_code=400, message=translate(lang, "validation.require_auth"))
    
    ### Check permissions ###
    access, message = check_permissions(db, request, "platform.roles.create")
    
    if not access:
        return custom_response(status_code=400, message=message)

    ### Operations ###
    permissions = get_permissions()
        
    return custom_response(status_code=200, message=translate(lang, "platform.roles.get.success"), data={
        "permissions": permissions
    })

########## Create Role - POST ##########
@router.post("/create")
async def create_role(request: Request, db: Session = Depends(get_db)):
    ### Variables ###
    lang = request.state.lang
    user = request.state.user

    ### Validation ###
    if user == None:
        return custom_response(status_code=400, message=translate(lang, "validation.require_auth"))
    
    ### Check permissions ###
    access, message = check_permissions(db, request, "platform.roles.create")
    
    if not access:
        return custom_response(status_code=400, message=message)
    
    ### Get Body ###
    new_role_data, error = await read_json_body(request)
    if error: 
        return custom_response(status_code=400, message=error)
        
    required_fields, error = validate_required_fields(new_role_data, ["permissions", "role_name", "description", "hidden"])
    if error:
        return custom_response(status_code=400, message=translate(lang, "validation.required_f"), details=required_fields)
        
    permissions = filter_existing_permissions(new_role_data.permissions)

    new_role = User_Role(
        id = get_uuid(db, User_Role),
        name = new_role_data.role_name,
        description = new_role_data.description,
        permissions = permissions,
        platform_level = True
    )

    if new_role_data.hidden == "1":
        new_role.hidden = True

    add_db(db, new_role)
        
    return custom_response(status_code=200, message=translate(lang, "platform.roles.create.success"), data={
        "role_id": new_role.id
    })

########## Get Role - To Update Role ##########
@router.get("/get/{role_id}")
async def get_roles(request: Request, role_id: str, db: Session = Depends(get_db)):
    ### Variables ###
    lang = request.state.lang
    user = request.state.user
    permissions = []

    ### Validation ###
    if user == None:
        return custom_response(status_code=400, message=translate(lang, "validation.require_auth"))
    
    ### Check permissions ###
    access, message = check_permissions(db, request, "platform.roles.update")
    
    if not access:
        return custom_response(status_code=400, message=message)
    
    if not role_id:
        return custom_response(status_code=400, message=translate(lang, "platform.roles.get.single.error"))
    
    check_role = db.query(User_Role).filter(
        User_Role.id == role_id
    ).first()
    
    if not check_role:
        return custom_response(status_code=400, message=translate(lang, "platform.roles.get.single.error"))
        
    ### Operations ###
    permissions = get_permissions()

    permissions_selected = {
        "name": check_role.name,
        "description": check_role.description,
        "permissions": check_role.permissions,
        "hidden": check_role.hidden
    }
        
    return custom_response(status_code=200, message=translate(lang, "platform.roles.get.single.success"), data={
        "permissions": permissions,
        "permissions_selected": permissions_selected
    })

########## Post Role - To Update Role ##########
@router.post("/update/{role_id}")
async def get_roles(request: Request, role_id: str, db: Session = Depends(get_db)):
    ### Variables ###
    lang = request.state.lang
    user = request.state.user
    permissions = []

    ### Validation ###
    if user == None:
        return custom_response(status_code=400, message=translate(lang, "validation.require_auth"))
    
    ### Check permissions ###
    access, message = check_permissions(db, request, "platform.roles.update")
    
    if not access:
        return custom_response(status_code=400, message=message)
    
    if not role_id:
        return custom_response(status_code=400, message=translate(lang, "platform.roles.get.single.error"))
    
    ### Get Body ###
    update_role, error = await read_json_body(request)
    if error: 
        return custom_response(status_code=400, message=error)

    required_fields, error = validate_required_fields(update_role, ["permissions", "role_name", "description", "hidden"])
    if error:
        return custom_response(status_code=400, message=translate(lang, "validation.required_f"), details=required_fields)

    check_role = db.query(User_Role).filter(
        User_Role.id == role_id
    ).first()

    if not check_role:
        return custom_response(status_code=400, message=translate(lang, "platform.roles.get.single.error"))
    
    permissions = filter_existing_permissions(update_role.permissions)
    print(permissions)

    check_role.name = update_role.role_name
    check_role.permissions = permissions
    check_role.description = update_role.description

    if update_role.hidden == "1":
        check_role.hidden = True
    else:
        check_role.hidden = False

    update_db(db)
        
    return custom_response(status_code=200, message=translate(lang, "platform.roles.update.single.success"), data={
        "role_id": check_role.id
    })

File: frontend/src/routes/platform/roles.js
/*************** Modules ***************/
import express from 'express';

import { get_data, send_data } from '../../utils/api.js';
import { require_auth, platform_mod } from '../../middlewares/auth.js';

/*************** Variables ***************/
const router = express.Router();

/*************** Main route ***************/
router.get("/", require_auth, platform_mod, async (req, res) => {
    // Variables
    const permissions = req.permissions;

    let roles = [];

    // Check permissions
    if (!permissions.includes("platform.roles.read")) return res.redirect("/system-alert/403");

    // Get Structure
    const response = await get_data("/platform/roles", {}, req);

    if (response.data.roles) roles = response.data.roles;

    // Render content
    return res.render("platform/roles/main", {
        roles: roles
    });
});

/*************** Create Role - GET ***************/
router.get("/create", require_auth, platform_mod, async (req, res) => {
    // Variables
    const permissions = req.permissions;

    let permissions_data = [];

    // Check permissions
    if (!permissions.includes("platform.roles.create")) return res.redirect("/system-alert/403");

    // Get Structure
    const response = await get_data("/platform/roles/get-permissions", {}, req);

    if (response.data.permissions) {
        permissions_data = response.data.permissions
    }

    // Render content
    return res.render("platform/roles/create", {
        permissions_data: permissions_data
    });
});

/*************** Create Role - POST ***************/
router.post("/create", require_auth, platform_mod, async (req, res) => {
    // Variables
    const permissions = req.permissions;

    let errors = [];

    let hidden_v = "0";
    let permissions_data_v = [];
    let description_s = "No hay descripción";

    // Check permissions
    if (!permissions.includes("platform.roles.create")) return res.redirect("/system-alert/403");

    // Get POST data
    const { permissions_data, role_name, description, hidden } = req.body;

    // Correction
    if (description) description_s = description;

    if (hidden == "on") hidden_v = "1";

    if (typeof(permissions_data) == "string") permissions_data_v = [permissions_data]
    else permissions_data_v = permissions_data

    const response = await send_data("/platform/roles/create", {}, {
        role_name,
        
        permissions: permissions_data_v,
        description: description_s,
        hidden: hidden_v
    }, req);

    if (response.error) {
        let permissions_data_v = []

        if (response.details.length <= 0) errors = [response.message];
        else errors = response.details;

        const response2 = await get_data("/platform/roles/get-permissions", {}, req);

        if (response2.data.permissions) {
            permissions_data_v = response.data.permissions
        }

        return res.render("platform/roles/create", {
            permissions_data: permissions_data_v
        });
    }

    const data = response.data;

    // Render content
    return res.redirect(`/platform/roles/view/${data.role_id}`);
});

/*************** Edit Role - GET ***************/
router.get("/update/:role_id", require_auth, platform_mod, async (req, res) => {
    // Variables
    const permissions = req.permissions;
    const role_id = req.params.role_id;

    let permissions_data = [];
    let permissions_selected_data = {};

    // Check permissions
    if (!permissions.includes("platform.roles.update")) return res.redirect("/system-alert/403");

    // Get Structure
    const response = await get_data(`/platform/roles/get/${role_id}`, {}, req);

    if (response.error) {
        return res.redirect("/platform/roles");
    }

    permissions_data = response.data.permissions
    permissions_selected_data = response.data.permissions_selected

    // Render content
    return res.render("platform/roles/update", {
        role_id: role_id, 
        permissions_data, 
        permissions_selected_data
    });
});

/*************** Edit Role - PUT ***************/
router.put("/update/:role_id", require_auth, platform_mod, async (req, res) => {
    // Variables
    const permissions = req.permissions;
    const role_id = req.params.role_id;

    let hidden_v = "0";
    let description_s = "No hay descripción";
    let permissions_data_v = [];

    // Check permissions
    if (!permissions.includes("platform.roles.update")) return res.redirect("/system-alert/403");

    // Get POST data
    const { permissions_data, role_name, description, hidden } = req.body;

    if (typeof(permissions_data) == "string") permissions_data_v = [permissions_data]
    else permissions_data_v = permissions_data

    // Correction
    if (description) description_s = description;

    if (hidden == "on") hidden_v = "1";

    // Operations
    const response = await send_data(`/platform/roles/update/${role_id}`, {}, {
        role_name, 

        permissions: permissions_data_v,
        description: description_s,
        hidden: hidden_v
    }, req);

    if (response.error) {
        const response2 = await get_data(`/platform/roles/get/${role_id}`, {}, req);

        const permissions_data = response2.data.permissions;
        const permissions_selected_data = response2.data.permissions_selected;

        if (response2.error) {
            return res.redirect("/platform/roles");
        }

        return res.render("platform/roles/update", {
            errors: [response.message], 
            permissions_data, permissions_selected_data
        });
    }

    // Render content
    return res.redirect(`/platform/roles/update/${role_id}`);
});

/*************** Export ***************/
export default router
